### ISSUE TITLE
Packer fails with unprivileged LXC containers

### ISSUE THREAD

**Issue by zhukovra** [hashicorp/packer#6781]
Hi all.

I'm trying to build LXC boxes but get an error on creating containers with packer: https://gist.github.com/zhukovra/20539596a86ac7ed20c60addf5b26c15

My system:
```
/t/lxc $ packer version
Packer v1.3.1
/t/lxc $ uname -a
Linux z12 4.15.0-34-generic #37~16.04.1-Ubuntu SMP Tue Aug 28 10:44:06 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

Creating container with LXC tools is working:
```
/t/lxc $ lxc-create -t download -n packer-lxc -- -d ubuntu -r trusty -a amd64
Using image from local cache
Unpacking the rootfs

---
You just created an Ubuntu trusty amd64 (20180928_07:42) container.

To enable SSH, run: apt install openssh-server
No default root or user password are set by LXC.
/t/lxc $ lxc-start -n packer-lxc -d
/t/lxc $ lxc-ls -f
NAME       STATE   AUTOSTART GROUPS IPV4       IPV6 
packer-lxc RUNNING 0         -      10.0.3.194 -   
```

**Comment by rickard-von-essen** [comment]
Duplicates #6545?

**Comment by zhukovra** [comment]
@rickard-von-essen I'm not sure, because packer fails on first step - creating container for provision/exporting.

**Comment by ChrisLundquist** [comment]
This *could* (I have not verified) be related to the lxc tools being upgraded. 
I wrote this against 16.04 which has:
https://packages.ubuntu.com/xenial/lxc1
(2.0.7-0ubuntu1~16.04.2 and others)

and 18.04 now has:
https://packages.ubuntu.com/bionic/lxc1
(3.0.1-0ubuntu1~18.04.2)

Do we have any established patterns for handling versioning like this for other builders?

**Comment by ChrisLundquist** [comment]
It has been ages since I looked at this.

TODO 
- I should update the docs about `lxc-template`. I spent 10 minutes because I didn't have the package installed, and it failed silently.
- We should pass `-l DEBUG` to `lxc-create` when we invoke packer in debug mode (it helped me solve the above)

It looks like the source of this issue is:
```
$  env SUITE=trusty lxc-create -n packer-lxc-trusty -t ubuntu -l DEBUG
lxc-create: packer-lxc-trusty: conf.c: chown_mapped_root: 3046 No uid mapping for container root
lxc-create: packer-lxc-trusty: lxccontainer.c: do_storage_create: 1275 Error chowning "/home/clundquist/.local/share/lxc/packer-lxc-trusty/rootfs" to container root
lxc-create: packer-lxc-trusty: conf.c: suggest_default_idmap: 4561 You must either run as root, or define uid mappings
lxc-create: packer-lxc-trusty: conf.c: suggest_default_idmap: 4562 To pass uid mappings to lxc-create, you could create
lxc-create: packer-lxc-trusty: conf.c: suggest_default_idmap: 4563 ~/.config/lxc/default.conf:
lxc-create: packer-lxc-trusty: conf.c: suggest_default_idmap: 4564 lxc.include = /etc/lxc/default.conf
lxc-create: packer-lxc-trusty: conf.c: suggest_default_idmap: 4565 lxc.idmap = u 0 165536 65536
lxc-create: packer-lxc-trusty: conf.c: suggest_default_idmap: 4566 lxc.idmap = g 0 165536 65536
lxc-create: packer-lxc-trusty: lxccontainer.c: do_lxcapi_create: 1830 Failed to create (none) storage for packer-lxc-trusty
Error creating container packer-lxc-trusty
```

**Comment by ChrisLundquist** [comment]
More notes:
- It appears that the invocation for `lxc-template/lxc-ubuntu` has changed to `-r trusty` rather than env vars?
- It appears unprivileged containers need to use the `download` template because they can't mknode, mountfs and things. see [ubuntu lxc help](https://help.ubuntu.com/lts/serverguide/lxc.html) and [linux containers getting started](https://linuxcontainers.org/lxc/getting-started/)

Locally, I'm stuck at about the same place as this https://github.com/lxc/lxc/issues/2461. `sudo` makes it work, but that isn't what we want.

If @zhukovra has unprivileged containers working locally, then he can use the download template here with packer. The format would be something like:
```
{
  "builders": [
    {
      "type": "lxc",
      "name": "lxc-trusty",
      "config_file": "/etc/lxc/default.conf",
      "template_name": "download",
      "template_parameters": [ "-r trusty" ]
    }
  ]
}
```

So, to recap:
- Unprivileged containers may never have worked, or don't work with 3.x
- Many of the deps have changed format in 3.x so our example file is out of date.
- We should add some more getting started docs.
- We should add some more debugging flags where we can.

**Comment by ChrisLundquist** [comment]
Burried deep in the logs @zhukovra provided:
https://gist.github.com/zhukovra/20539596a86ac7ed20c60addf5b26c15#file-packer-output-L111-L122

It says use the download template for unprivileged containers.
```
2018/09/28 16:06:44 packer: 2018/09/28 16:06:44 stderr: This template can't be used for unprivileged containers.
2018/09/28 16:06:44 packer: You may want to try the "download" template instead.
2018/09/28 16:06:44 packer: lxc-create: lxccontainer.c: create_run_template: 1295 container creation template for packer-lxc-trusty failed
2018/09/28 16:06:44 packer: lxc-create: tools/lxc_create.c: main: 318 Error creating container packer-lxc-trusty
2018/09/28 16:06:44 ui error: ==> lxc-trusty: Error creating container: Command error: This template can't be used for unprivileged containers.
==> lxc-trusty: You may want to try the "download" template instead.
```

**Comment by zhukovra** [comment]
Hi, @ChrisLundquist! Thank you for investigation, I can create unprivileged containers:
```
$ lxc-create -t download -n packer-1 -- -r jessie -d debian -a amd64
Using image from local cache
Unpacking the rootfs
---
You just created a Debian jessie amd64 (20180928_05:24) container.

To enable SSH, run: apt install openssh-server
No default root or user password are set by LXC.
$ lxc-destroy --name packer-1
Destroyed container packer-1
```

But your suggested config doesn't work:
```
{
  "builders": [
    {
      "type": "lxc",
      "name": "lxc-jessie-1",
      "config_file": "/etc/lxc/default.conf",
      "template_name": "download",
      "template_parameters": [ "-r jessie", "-d debian", "-a amd64" ]
    }
  ]
}
```
[Output](https://gist.github.com/zhukovra/20539596a86ac7ed20c60addf5b26c15#file-output-2-txt) (output-2.txt). There is the error:
```
2018/10/09 15:41:07 packer: 2018/10/09 15:41:07 Executing args: []string{"env", "lxc-create", "-n", "packer-lxc-jessie-1", "-t", "download", "--", "-r jessie", "-d debian", "-a amd64"}
2018/10/09 15:41:10 packer: 2018/10/09 15:41:10 stdout: Setting up the GPG keyring
2018/10/09 15:41:10 packer: Downloading the image index
2018/10/09 15:41:10 packer: ERROR: Couldn't find a matching image.
```
I don't know why executing similar `lxc-create` command from packer can't create container.

**Comment by ChrisLundquist** [comment]
Try splitting “-r Debian” into “-r”, “Debian”, and the same for the other
two strings. Exec is passing it as one token and that could break arg
parsing.

**Comment by milgelren** [comment]
Hi,
As there hasn’t been any recent activity on this issue, we’re closing it for now. If the issue persists or you have new information to share, feel free to reopen it or create a new one.

If you found a solution elsewhere or received help through the [community forum](https://discuss.hashicorp.com/c/packer), we’d appreciate it if you could share an update here for others who might encounter the same problem.

Thanks!
